Allow to prevent base64 conversions by using bytes

spinscale opened the pull request:
CBOR is natively supported in Elasticsearch and allows for byte arrays.
This means, that by using CBOR the user can prevent base64 conversions
for the data being sent back and forth.

This PR adds support to extract data from a byte array in addition to
a string.

javanna commented on plugins/ingest-attachment/src/test/java/org/elasticsearch/ingest/attachment/AttachmentProcessorTests.java:
```diff
@@ -168,11 +166,48 @@ public void testEpubDocument() throws Exception {
+        document.put("source_field", randomObject);
+        IngestDocument ingestDocument = RandomDocumentPicks.randomIngestDocument(random(), document);
+        try {
+            processor.execute(ingestDocument);
```
missing fail() here?

javanna commented:
left a minor comment LGTM otherwise

martijnvg commented:
LGTM

kimchy commented:
maybe we can have special support for `IngestDocument#getFieldValue(String path, byte[].class)`? then we can have that logic encapsulated in IngestDocument and used in other processors down the road?

kimchy commented:
btw, if we do that, we can only test `IngestDocument` in a unit test that it handles the `byte[]` case correctly, and thats it. No need I think to add an integration test (we shouldn't test CBOR support in ES, we already do that elsewhere) for example.

spinscale commented:
that makes sense. I added a `IngestDocument.getFieldValueAsBytes()` method, which is doing the base64 conversion, instead of doing this in the `AttachmentProcessor`.

Also removed the integration test

javanna commented on core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java:
```diff
@@ -111,6 +115,36 @@ public IngestDocument(Map<String, Object> sourceAndMetadata, Map<String, String>
         return cast(path, context, clazz);
     }
 
+    /**]
```
additional square bracket :)

javanna commented on core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java:
```diff
@@ -111,6 +115,36 @@ public IngestDocument(Map<String, Object> sourceAndMetadata, Map<String, String>
+        Object context = fieldPath.initialContext;
+        for (String pathElement : fieldPath.pathElements) {
+            context = resolve(pathElement, path, context);
+        }
```
does it make sense to call getFieldValue(path, Object.class) here instead ?

javanna commented on core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java:
```diff
@@ -111,6 +115,36 @@ public IngestDocument(Map<String, Object> sourceAndMetadata, Map<String, String>
+            throw new IllegalArgumentException("Content field [" + path + "] of unknown type [" + context.getClass().getName() +
+                "], must be string or byte array");
+        }
+
```
remove this additional line break? :)

javanna commented on core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java:
```diff
@@ -492,7 +526,7 @@ private static Object deepCopy(Object value) {
         } else if (value == null || value instanceof String || value instanceof Integer ||
             value instanceof Long || value instanceof Float ||
-            value instanceof Double || value instanceof Boolean) {
+            value instanceof Double || value instanceof Boolean || value instanceof byte[]) {
```
I think that we need to deep copy the byte array instead of returning it as-is. Otherwise subsequent changes to the array against a copy of the document will affect all the other copies too.

javanna commented on plugins/ingest-attachment/src/test/java/org/elasticsearch/ingest/attachment/AttachmentProcessorTests.java:
```diff
@@ -57,8 +57,7 @@ public void createStandardProcessor() throws IOException {
     public void testEnglishTextDocument() throws Exception {
         Map<String, Object> attachmentData = parseDocument("text-in-english.txt", processor);
 
```
does it make sense to add some test for the new getFieldValueAsBytes (or maybe move some of this below) to IngestDocumentTests ?

spinscale commented:
I just found out that using byte arrays and copying them breaks the equality check in the tests... the bytesreference workaround I had in mind is broken. Will revert this and come up with something more useful.

martijnvg commented on core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java:
```diff
@@ -112,6 +118,32 @@ public IngestDocument(Map<String, Object> sourceAndMetadata, Map<String, String>
+        Object object = getFieldValue(path, Object.class);
+        if (object instanceof byte[]) {
+            byte[] bytes = (byte[]) object;
+            return Arrays.copyOf(bytes, bytes.length);
```
why do we need to copy the bytes here?

spinscale replied:
we absolutely dont.. my bad

martijnvg commented:
LGTM

javanna commented:
LGTM as well

clintongormley commented:
@spinscale want to get this in?

spinscale commented:
this one still has breaking tests that need to be fixed first

spinscale commented:
@martijnvg I finally resurrected this one and fixed the outstanding tests. The main issue was parts of the code rely on equality, which requires a custom `equals` implementation for byte arrays. The same applies for the `ValueSource` interface

What I did here, was adding changing `equals` in `IngestDocument` so that byte arrays are converted to an `ByteArray` object that uses `Arrays.equals()` in its `equals` method. The other possibility would be to do this conversion in the ctor when the `IngestDocument` is instantiated, but I felt it makes more sense to defer only when `equals` is called. Maybe you have a better idea that I overlooked, happy to get any input.

martijnvg commented on core/src/test/java/org/elasticsearch/ingest/core/IngestDocumentTests.java:
```diff
@@ -969,31 +970,24 @@ public void testIngestMetadataTimestamp() throws Exception {
     public void testCopyConstructor() {
         IngestDocument ingestDocument = RandomDocumentPicks.randomIngestDocument(random());
         IngestDocument copy = new IngestDocument(ingestDocument);
-        recursiveEqualsButNotSameCheck(ingestDocument.getSourceAndMetadata(), copy.getSourceAndMetadata());
```
Only this test was failing, right?
Why not just change the `recursiveEqualsButNotSameCheck(...)` method to do `arrayEquals(...)` in case of arrays?

martijnvg commented:
LGTM
